invenio_records_permissions/generators.py

```diff
@@ -6,227 +6,192 @@
+        """Constructor."""
+        super(Admin, self).__init__()
 
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
+        return [ActionNeed('admin-access')]
 
-class _RecordOwners(_RecordNeedClass):
 
-    def __init__(self):
-        super(_RecordOwners, self).__init__()
+class RecordOwners(Generator):
+    """Allows record owners."""
 
-    def needs(self, record):
-        owner_needs = []
-        for owner in record.get('owners', []):
-            owner_needs.append(UserNeed(owner))
-        return owner_needs
+    def needs(self, record=None, **kwargs):
+        """Enabling low-level Needs."""
+        return [UserNeed(owner) for owner in record.get('owners', [])]
 
-    def query_filter(self):
+    def query_filter(self, record=None, **kwargs):
+        """Filters for current identity as owner."""
         provides = g.identity.provides
         for need in provides:
             if need.method == 'id':
                 return Q('term', owners=need.value)
-        return None
-
-
-RecordOwners = _RecordOwners()
-
-
-class _DepositOwners(_RecordNeedClass):
+        return []
 
-    def __init__(self):
-        super(_DepositOwners, self).__init__()
-
-    def needs(self, record):
-        deposit_owners = []
-        for owner in record.get('deposits', {}).get('owners', []):
-            deposit_owners.append(UserNeed(owner))
-        return deposit_owners
-
-    def query_filter(self):
-        provides = g.identity.provides
-        for need in provides:
-            if need.method == 'id':
-                return Q('term', **{"deposits.owners": need.value})
-        return None
 
+class AnyUserIfPublic(Generator):
+    """Allows any user if record is public."""
 
-DepositOwners = _DepositOwners()
+    def needs(self, record=None, **rest_over):
+        """Enabling low-level Needs."""
+        is_restricted = (
+            record and
+            record.get('_access', {}).get('metadata_restricted', False)
+        )
+        return [any_user] if not is_restricted else []
 
-
-class IfPublicRecordFactory(_RecordNeedClass):
-
-    def __init__(self, is_restricted, es_filter):
-        super(IfPublicRecordFactory, self).__init__()
-        self.is_restricted = is_restricted
-        self.es_filter = es_filter
-
-    def needs(self, record, *args, **kwargs):
-        if not self.is_restricted(record, *args, **kwargs):
-            return [any_user]
-        else:
-            return None
-
-    def excludes(self, record, *args, **kwargs):
-        pass
+    def excludes(self, record=None, **rest_over):
+        """Preventing low-level Needs."""
+        return []
 
     def query_filter(self, *args, **kwargs):
-        return self.es_filter(*args, **kwargs)
+        """Filters for non-restricted records."""
+        return Q('term', **{"_access.metadata_restricted": False})
 
 
-def _is_restricted(record, *args, **kwargs):
-    if record:
-        # FIXME: this should be caster to boolean when loaded
```

Comment:
Thanks for fixing this! Not sure what pass by my head by putting `json.loads`...

Reply:
:+1: